pkg/util/helper/work.go

```diff
@@ -83,6 +84,39 @@ func GetWorksByLabelsSet(c client.Client, ls labels.Set) (*workv1alpha1.WorkList
 	return workList, c.List(context.TODO(), workList, listOpt)
 }
 
+// GetWorksByBindingNamespaceName get WorkList by matching same Namespace and same Name.
+func GetWorksByBindingNamespaceName(c client.Client, bindingNamespace, bindingName string) (*workv1alpha1.WorkList, error) {
+	referenceKey := names.GenerateBindingReferenceKey(bindingNamespace, bindingName)
+	var ls labels.Set
+	if bindingNamespace != "" {
+		ls = labels.Set{workv1alpha2.ResourceBindingReferenceKey: referenceKey}
+	} else {
+		ls = labels.Set{workv1alpha2.ClusterResourceBindingReferenceKey: referenceKey}
+	}
+
+	workList, err := GetWorksByLabelsSet(c, ls)
+	if err != nil {
+		return nil, err
+	}
+	retWorkList := &workv1alpha1.WorkList{}
+	for _, work := range workList.Items {
+		var sameResourceBinding bool
+		if bindingNamespace != "" {
+			sameResourceBinding = util.GetAnnotationValue(work.GetAnnotations(), workv1alpha2.ResourceBindingNameAnnotationKey) == bindingName &&
+				util.GetAnnotationValue(work.GetAnnotations(), workv1alpha2.ResourceBindingNamespaceAnnotationKey) == bindingNamespace
+		} else {
+			sameResourceBinding = util.GetAnnotationValue(work.GetAnnotations(), workv1alpha2.ClusterResourceBindingAnnotationKey) == bindingName
+		}
+		// If the annotation is different from ResourceBinding, ignore it.
+		// refer issue https://github.com/karmada-io/karmada/issues/2071
+		if sameResourceBinding {
+			retWorkList.Items = append(retWorkList.Items, work)
+		}
+	}
```

Comment:
```suggestion
	// Due to the hash collision problem, we have to filter the Works by annotation.
	// More details please refer to https://github.com/karmada-io/karmada/issues/2071.
	for i := range workList.Items {
		if len(bindingNamespace) > 0 { // filter Works that derived by 'ResourceBinding'
			if util.GetAnnotationValue(workList.Items[i].GetAnnotations(), workv1alpha2.ResourceBindingNameAnnotationKey) == bindingName &&
				util.GetAnnotationValue(workList.Items[i].GetAnnotations(), workv1alpha2.ResourceBindingNamespaceAnnotationKey) == bindingNamespace {
				retWorkList.Items = append(retWorkList.Items, workList.Items[i])
			}
		} else { // filter Works that derived by 'ClusterResourceBinding'
			if util.GetAnnotationValue(workList.Items[i].GetAnnotations(), workv1alpha2.ClusterResourceBindingAnnotationKey) == bindingName {
				retWorkList.Items = append(retWorkList.Items, workList.Items[i])
			}
		}
	}
```

I guess the `sameResourceBinding` is not accurate enough, my suggestion is just to remove it.
And during the process of iteration, I prefer using the `index` to avoid the memory copy.
Otherwise looks good to me.

Reply:
nice!